clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -352,6 +332,44 @@ else if (!future.isRetriable())
         }
     }
 
+    private synchronized void resetJoinGroupFuture() {
+        this.joinFuture = null;
```

Comment:
Where could `joinFuture` be accessed by the heartbeat thread? Do we need to guard it?